### DISCUSSION TITLE
Extract parser and publish it as a separate crate

### DISCUSSION THREAD

**Discussion by ngirard** [zoni/obsidian-export#62] [Feature requests]
Hi,
let me start by expressing my gratitude for your efforts into this nice project !

Your project is the only one (in Rust) I know of that parses Obsidian files properly.
It would be great if you made the parser available to other projects (including the one I'm trying to implement as I'm writing) !
Skimming through your code, that would be the whole contents of `references.rs`, plus the YAML and Markdown parsing logic embedded within `Exporter::run()` from `lib.rs`. I might have missed something, but I guess you'll know better ;-)

Cheers

**Comment by zoni** [comment]
Hey @ngirard,

Doing this has crossed my mind in the past as well. While it's not as trivial as simply extracting that code and publishing it separately, the refactoring I've done in the last 2 or so versions has made it easier to do something like this.

Partly as a note-to-self, some of the things that should happen for this:

- We should extract [parse_obsidian_note](https://github.com/zoni/obsidian-export/blob/5ff990ca2054b32c47d961f92c82537d8c9a16fd/src/lib.rs#L394-L398). This should probably become the main function people would call from this new library/crate.
- [context](https://github.com/zoni/obsidian-export/blob/5ff990ca2054b32c47d961f92c82537d8c9a16fd/src/lib.rs#L397) should be removed from the interface, I think that only makes sense for the recursive parsing obsidian-export does itself. This requires moving the infinite recursion logic elsewhere though (https://github.com/zoni/obsidian-export/blob/5ff990ca2054b32c47d961f92c82537d8c9a16fd/src/lib.rs#L399-L402).
- This function should then stop resolving embedded notes itself (https://github.com/zoni/obsidian-export/blob/5ff990ca2054b32c47d961f92c82537d8c9a16fd/src/lib.rs#L489-L497). Instead it should return some sort of "Here's an embed to note foo" element. This requires changing the return type to some sort of enum that can distinguish between regular markdown events and special obsidian tags.

I'm unsure how I feel about immediately publishing this logic as a separate crate, partly because most of the tests for the parsing logic are integration tests which rely on more than just this piece of code and so cannot be lifted over without changes.

Would you accept having the entire `obsidian-export` crate be a dependency on your project, if all of the parsing logic was in a separate module you could use independently of the rest? It's obviously non-ideal because you'd be pulling in more dependencies than strictly necessary, but would let you use my parser without me immediately having to refactor all the tests or needing to keep two crate releases in sync.

**Comment by ngirard** [comment]
@zoni, thanks for your reply and apologizes for being slow to respond, I'm being a bit overwhelmed with my work.

> I'm unsure how I feel about immediately publishing this logic as a separate crate, partly because most of the tests for the parsing logic are integration tests which rely on more than just this piece of code and so cannot be lifted over without changes.

Please take your time, there's no rush here on my side.

> Would you accept having the entire obsidian-export crate be a dependency on your project, if all of the parsing logic was in a separate module you could use independently of the rest? It's obviously non-ideal because you'd be pulling in more dependencies than strictly necessary, but would let you use my parser without me immediately having to refactor all the tests or needing to keep two crate releases in sync.

My project has different needs than Obsidian-export (I'm aiming at watching for file changes within a vault and maintaining an index in an SQLite database), so based on my first look to your code, even if you made quick arrangements for me, I'm not even sure I could benefit from them immediately.

I'd suggest taking the long run at your own pace, refactoring your code when you can. It wouldn't be a problem at all to me if you needed several months before publishing your parser as a separate crate. I'll celebrate when it's out !

Cheers
